mlir/lib/Gradient/IR/GradientOps.cpp

```diff
@@ -57,12 +57,19 @@ LogicalResult verifyGradInputs(OpState *op_state, func::FuncOp callee, ValueRang
                << "expected " << fnType.getNumInputs() << " but got " << fnArgs.size();
 
     if (callee->getAttrOfType<UnitAttr>(catalyst::quantum::hasInvalidGradientOp)) {
-        return op_state->emitOpError("An operation without a valid gradient was found in code "
-                                     "reachable from the gradient operation.\n"
-                                     "Example of operations not allowed:\n"
-                                     " * mid circuit measurements\n"
-                                     " * callbacks\n"
-                                     " * ZNE mitigation.");
+        if (dyn_cast<GradOp>(op_state->getOperation()) != nullptr) {
+            if (dyn_cast<GradOp>(op_state->getOperation()).getMethod() != "fd") {
```

Comment:
Give this a try. I don't think you need the comparison with `nullptr`. And you can also assign the expression to a variable.

```suggestion
        if (auto gradOp = dyn_cast<GradOp>(op_state->getOperation())) {
            if (gradOp.getMethod() != "fd") {
```

Reply:
By the way, c++ does indeed shortcut logical evaluations: 
```C++
#include <iostream>
#include <vector>

bool f(std::vector<int> *x){
   x->push_back(10);
   return false;
} 

int main(){
   std::vector<int> *x = nullptr;
   //f(x);   //segfault
   if (true || f(x)){
      std::cout << "x is ten!\n";
   }
}

>>>
x is ten!
```